docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+
+#### odo deploy 
+This command will build a container image for the application and deploy it on the target Kubernetes environment. 
+
+Flags:
+ - `--tag`: The tag to be used for the built application container image - `<registry>/<org>/<name>:<tag>` (mandatory).
+ - `service-account`: The service account for running privileged containers and push access to the image registry.
+ - `--credentials`: The credentials needed to push the image to the container image registry (mandatory).
+
+#### odo deploy delete
+This command will delete any resources created by odo deploy.
+
+### Detailed design: 
+
+### Devfile
+
+For the initial implementation, we could use devfile v2.0.0 and capture basic outer-loop information as `attributes` on the devfile. `odo` could look for  specific keys, while other tools like Che could ignore them.
+
+For example: 
+```
+attributes:
+    build-dockerfile: <URI>
+    deployment-manifest: <URI>
+```
+
+This should not need any change to the devfile v2 schema or the parser.
+
+### Dockerfile
+This could be any valid dockerfile.
+
+### Deployment Manifest
+The deployment manifest could be templated to help with replacing key bits of information:
+- PROJECT_NAME
+- CONTAINER_IMAGE
+- PORT
+
+For example: 
+```
+apiVersion: app.stacks/v1beta1
+kind: RuntimeComponent
+metadata:
+    name: PROJECT_NAME
+spec:
+    applicationImage: CONTAINER_IMAGE 
+    service:
+        type: ClusterIP
+        port: PORT
+    expose: true
+    storage:
+        size: 2Gi
+        mountPath: "/logs"
+```
+
+### odo deploy
+This command will perform the following actions:
+
+#### Input validation
+- Check if the devfile  is v2.0.0 and that it specifies the expected outer-loop attributes. 
+    - If not provided, display a meaningful error message.
+- Validate all arguments passed by the user. 
+    - If argument values are invalid, display a meaningful error message.
+
+#### Build
+- Start a new pod with a build container (privileged container with `buildah:stable` image).
+- Mount the image registry credentials passed by the user.
+- Copy the application source code into the build container.
+- Fetch the dockerfile using URI in the attributes of the devfile and make it available in the container.
+- Execute `buildah bud` command using src code, dockerfile and tag information.
+- Push the built image using the tag and credentials.
+- Delete the build container and the pod.
+
+#### Deploy
+- Delete any existing deployments (if the user runs `odo deploy` multiple times).
+- Fetch the deployment manifest using URI in the attributes of the devfile. and make it available in the container.
+- Replace templated text in the deployment manifest with relevant values:
+    - PROJECT_NAME: name of odo project
+    - CONTAINER_IMAGE: `tag` for the built image
+    - PORT: URL information in env.yaml
+- Apply the new deployment manifest.
+- Save the deployment manifest in `.odo` folder.
+- Provide the user with a URL for accessing the deployed application. 
```

Comment:
`odo deploy` automagically creates a URL without having to use `odo url create`? Where does it get the info to create URL from? Sorry if I'm missing something from devfile side of the discussion on this.

Reply:
The initial idea was to expect user has done `odo url create` during development so we know key information about the PORT etc. Rest of the bits about route should be handled/provided by the deployment manifest. 

In the example of runtime component operator listed above, the operator reads the CR, and then creates necessary routes etc. If that is not the case then the route should be part of the deployment manifest too with placeholder values for app specific info.

@groeges  please correct me if I got this wrong.

Reply:
The initial idea was to expect user has done `odo url create` during development so we know key information about the PORT etc. Rest of the bits about route should be handled/provided by the deployment manifest. 

In the example of runtime component operator listed above, the operator reads the CR, and then creates necessary routes etc. If that is not the case then the route should be part of the deployment manifest too with placeholder values for app specific info.

@groeges  please correct me if I got this wrong.

Reply:
Yes, the proposal was that the `odo url create` had already been done and the `odo deploy` would use that info to substitute the relevant PORT info into the manifest (if these template variables had been included). If the manifest wasn't configured to produce a Route/Endpoint, ie explicitly in the manifest or via an Operator then one would not be automatically created and no URL would be returned.   

This is one of the benefits of testing the deployment using `odo deploy` in order to test that the application is suitable to be deployed using the manifest etc that will be used in full outer-loop support with pipelines etc.

Reply:
Makes sense. Thanks for the explanation @neeraj-laad & @groeges :+1: